### PR TITLE
Emit special instruction for array literal + .(hash|min|max)

### PR DESCRIPTION
**Pull request by tenderlove**
This commit introduces a new instruction `opt_newarray_send` which is
used when there is an array literal followed by either the `hash`,
`min`, or `max` method.

```
[a, b, c].hash
```

Will emit an `opt_newarray_send` instruction.  This instruction falls
back to a method call if the "interested" method has been monkey
patched.

Here are some examples of the instructions generated:

```
$ ./miniruby --dump=insns -e '[@a, @b].max'
== disasm: #<ISeq:<main>@-e:1 (1,0)-(1,12)> (catch: FALSE)
0000 getinstancevariable                    :@a, <is:0>               (   1)[Li]
0003 getinstancevariable                    :@b, <is:1>
0006 opt_newarray_send                      2, :max
0009 leave
$ ./miniruby --dump=insns -e '[@a, @b].min'
== disasm: #<ISeq:<main>@-e:1 (1,0)-(1,12)> (catch: FALSE)
0000 getinstancevariable                    :@a, <is:0>               (   1)[Li]
0003 getinstancevariable                    :@b, <is:1>
0006 opt_newarray_send                      2, :min
0009 leave
$ ./miniruby --dump=insns -e '[@a, @b].hash'
== disasm: #<ISeq:<main>@-e:1 (1,0)-(1,13)> (catch: FALSE)
0000 getinstancevariable                    :@a, <is:0>               (   1)[Li]
0003 getinstancevariable                    :@b, <is:1>
0006 opt_newarray_send                      2, :hash
0009 leave
```

Co-authored-by: John Hawthorn <jhawthorn@github.com>

### CODE REVIEW

File: compile.c [hunk 1]
```diff
@@ -3525,13 +3525,18 @@ iseq_specialized_instruction(rb_iseq_t *iseq, INSN *iobj)
...
+                      {
+                          rb_num_t num = (rb_num_t)iobj->operands[0];
+                          iobj->insn_id = BIN(opt_newarray_send);
+                          iobj->operands = compile_data_calloc2(iseq, insn_len(iobj->insn_id) - 1, sizeof(VALUE));
+                          iobj->operands[0] = (VALUE)num;
+                          iobj->operands[1] = (VALUE)rb_id2sym(vm_ci_mid(ci));
+                          iobj->operand_size = insn_len(iobj->insn_id) - 1;
+                          ELEM_REMOVE(&niobj->link);
+                          return COMPILE_OK;
+                      }
```
Comment (nobu) [inline compile.c:3539]:
```suggestion
		  case idHash: {
                    rb_num_t num = (rb_num_t)iobj->operands[0];
                    iobj->insn_id = BIN(opt_newarray_send);
                    iobj->operands = compile_data_calloc2(iseq, insn_len(iobj->insn_id) - 1, sizeof(VALUE));
                    iobj->operands[0] = (VALUE)num;
                    iobj->operands[1] = rb_id2sym(vm_ci_mid(ci));
                    iobj->operand_size = insn_len(iobj->insn_id) - 1;
                    ELEM_REMOVE(&niobj->link);
                    return COMPILE_OK;
                  }
```

**Comment by nobu** [comment]
BTW, why can't we stop updating the autogenerated cruby_bindings.inc.rs manually?

**Comment by maximecb** [comment]
> BTW, why can't we stop updating the autogenerated cruby_bindings.inc.rs manually?

It's because constants like tHash and tFreeze were changed by this PR.

**Comment by tenderlove** [comment]
> > BTW, why can't we stop updating the autogenerated cruby_bindings.inc.rs manually?
> 
> It's because constants like tHash and tFreeze were changed by this PR.

I think @nobu is asking why we have to update `cruby_bindings.inc.rs` manually rather than have an automatic process do it.

**Comment by maximecb** [comment]
> > > BTW, why can't we stop updating the autogenerated cruby_bindings.inc.rs manually?
> > 
> > 
> > It's because constants like tHash and tFreeze were changed by this PR.
> 
> I think @nobu is asking why we have to update `cruby_bindings.inc.rs` manually rather than have an automatic process do it.

I think this is because @XrXr wanted to avoid forcing everyone to install the rust toolchain to build CRuby, but there might be an argument to be made that they already often have to in order to get the CI green?

**Review by maximecb — APPROVED** [review]

**Comment by XrXr** [comment]
> > > > BTW, why can't we stop updating the autogenerated cruby_bindings.inc.rs manually?
> > > 
> > > 
> > > It's because constants like tHash and tFreeze were changed by this PR.
> > 
> > 
> > I think @nobu is asking why we have to update `cruby_bindings.inc.rs` manually rather than have an automatic process do it.
> 
> I think this is because @XrXr wanted to avoid forcing everyone to install the rust toolchain to build CRuby, but there might be an argument to be made that they already often have to in order to get the CI green?

To update the file libclang/clang is also required but we need to support systems that only have GCC+rustc. An alternative is to generate the file when building the release tarball, but it's scary to add to that process because it's less visible and not as frequently exercised. CI for tarball builds is in a [different repo](https://github.com/ruby/actions) and breaking it can hold up releases.

**Comment by tenderlove** [comment]
Redmine feature is [here](https://bugs.ruby-lang.org/issues/18897)

**Pull request merged**